### PR TITLE
Support sharding serialized layers across ipus

### PR DESCRIPTION
**Pull request by kundaMwiza**
# What does this PR do?

- Adding the necessary IPU configuration options that will enable sharding serialized layers like `SerializedEmbedding` and `SplitLinear` across IPUs by using a pipeline configuration similar to `layers_per_ipu`. For example `[0, 0, 1, 2]` as an argument to `serialized_projection_splits_per_ipu` will result in 1 shard of a linear layer being placed on IPU2 and the last 2 shards on the last IPU.  The pipelines will also be split in a similar way for generation. For example if `layers_per_ipu=[0, 2, 2, 0]`, the corresponding encoder and decoder pipelines will be `[0, 2]` and `[2, 0]`. Similarly, `serialized_projection_layers_per_ipu` will be `[0, 0]` and `[1, 2]` for the encoder and decoder respectively. However the encoder portion `[0, 0]` is overrided with `None` as it is no longer a valid value for `serialized_projection_layers_per_ipu`. This behaviour is also similar for `SerializedEmbedding` and `serialized_embedding_splits_per_ipu` 
- Added `projection_serialization_factor` as a new parameter than can be used to serialize the LM head linear layers via `SerializedLinear` or `SplitProjection`. Some architectures rely on the use of `embedding_serialization_factor` for serializing linear layers, however that differs from the documented use case of `embedding_serialization_factor` in the `IPUConfig`. 

This feature will be used in the the `parallelize` method of a model:

```
if ipu_config._serialized_projection_splits_per_ipu:
     serialization_factor = sum(ipu_config._serialized_projection_splits_per_ipu)
     lm_head = SplitProjection(self.lm_head,  serialization_factor)
     lm_head.parallelize(ipu_config._serialized_projection_splits_per_ipu)
```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/ipu_configuration.py [hunk 1]
```diff
@@ -178,6 +209,74 @@ def __init__(self, **kwargs):
                 matmul_proportion = 0.2
             setattr(self, f"{mode}_matmul_proportion", kwargs.pop(f"{mode}_matmul_proportion", matmul_proportion))
 
+            # mode based layer configurations
+            # serialized_linear_splits_per_ipu and serialized_embedding_splits_per_ipu defaults
+            # depend on the model architecture
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:214]: These defaults don't seem to depend on the model architecture? Should this commend be moved to the validation loop below?

File: optimum/graphcore/ipu_configuration.py [hunk 2]
```diff
@@ -178,6 +209,74 @@ def __init__(self, **kwargs):
...
+                mode_layer_config = f"{mode}_serialized_{layer}_splits_per_ipu"
+                mode_layer_splits_per_ipu = getattr(self, mode_layer_config)
+                mode_layer_serialization_factor = getattr(self, f"{mode}_{layer}_serialization_factor")
+                if mode_layer_splits_per_ipu is not None:
+                    # serialized_{layer}_splits_per_ipu must be of type List[int>=0]
+                    if (
+                        not isinstance(mode_layer_splits_per_ipu, list)
+                        or not all(isinstance(x, int) for x in mode_layer_splits_per_ipu)
+                        or any(x < 0 for x in mode_layer_splits_per_ipu)
+                        or sum(mode_layer_splits_per_ipu) < 1
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:248]: This might just be personal preference but I find this version of the condition easier to read:
```suggestion
                    if not (
                        isinstance(mode_layer_splits_per_ipu, list)
                        and all(isinstance(x, int) for x in mode_layer_splits_per_ipu)
                        and all(x >= 0 for x in mode_layer_splits_per_ipu)
                        and sum(mode_layer_splits_per_ipu) > 0
```

File: optimum/graphcore/ipu_configuration.py [hunk 3]
```diff
@@ -190,36 +230,68 @@ def contents_geq_value_validator(
...
         if value not in allowed_values:
             raise ValueError(
                 f"`IPUConfig` attribute `output_mode` can only take values in"
                 f" {allowed_values}. You provided: {value=}"
             )
 
-    attribute_validators[output_mode_validator] = {"output_mode"}
+    attribute_validators[_output_mode_validator] = {"output_mode"}
+
+    def _serialized_layer_splits_per_ipu_validator(name: str, value: str):
```
Comment (katalinic-gc) [inline optimum/graphcore/ipu_configuration.py:267]: incorrect type hint for `value`

File: optimum/graphcore/modeling_utils.py [hunk 4]
```diff
@@ -509,6 +554,90 @@ def forward(self, x):
...
+        self,
+        linear: torch.nn.Linear,
+        serialization_factor: int,
+    ) -> None:
+        super().__init__()
+
+        self.in_features = linear.in_features
+        self.out_features = linear.out_features
+        if self.in_features % serialization_factor != 0:
+            raise ValueError(f"{linear.in_features=} must be divisible by {serialization_factor=}")
```
Comment (katalinic-gc) [inline optimum/graphcore/modeling_utils.py:580]: is this error msg correct?

File: optimum/graphcore/modeling_utils.py [hunk 5]
```diff
@@ -509,6 +554,90 @@ def forward(self, x):
...
+
+        self.split_size = int(self.in_features / serialization_factor)
```
Comment (katalinic-gc) [inline optimum/graphcore/modeling_utils.py:582]: if both of these are int, can you just do `self.in_features // serialization_factor`?

File: optimum/graphcore/modeling_utils.py [hunk 6]
```diff
@@ -509,6 +554,90 @@ def forward(self, x):
...
+        if linear.bias is not None:
+            self.bias = torch.nn.Parameter(torch.zeros_like(linear.bias))
+            with torch.no_grad():
+                self.bias.copy_(linear.bias.detach())
+
+    def forward(self, x):
+        # each linear layer processes a partition of the input
+        out = None
+        for i, split_linear_layer in enumerate(self.split_linear_layers):
+            h = split_linear_layer(x[:, :, i * self.split_size : (i + 1) * self.split_size])
```
Comment (katalinic-gc) [inline optimum/graphcore/modeling_utils.py:601]: I think this should be changed to `x[..., i * self.split_size ...]` instead of `x[:, :, i * self.split_size ... ]`, x could be 2D as well?

File: optimum/graphcore/modeling_utils.py [hunk 7]
```diff
@@ -509,6 +554,90 @@ def forward(self, x):
         return output
 
 
+class SplitProjection(torch.nn.Module):
```
Comment (jimypbr) [inline optimum/graphcore/modeling_utils.py:557]: Can you add a test that this gets same result as Linear?

File: tests/test_ipu_configuration.py [hunk 8]
```diff
@@ -206,41 +206,84 @@ def test_layers_per_ipu(self):
         self.assertEqual(layer_ipu, [0, 1, 1, 2, 2, 3, 3])
 
     def test_execution_mode_specific_options(self):
```
Comment (jimypbr) [inline tests/test_ipu_configuration.py:208]: this test class has ballooned in size. This needs to be broken up into different classes and smaller methods.

File: optimum/graphcore/modeling_utils.py (hunk 4 above)
Comment (kundaMwiza) [reply optimum/graphcore/modeling_utils.py:612]: Yeah, the serialisation/splitting is along the reduction dimension

File: tests/test_modeling_utils.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
+import unittest
+from abc import abstractmethod
+
+import pytest
+import torch
+
+import poptorch
```
Comment (kundaMwiza) [inline tests/test_modeling_utils.py:7]: @jimypbr added to test the serialized layers. Perhaps we have something like this elsewhere?

File: optimum/graphcore/modeling_utils.py [hunk 10]
```diff
@@ -541,6 +586,93 @@ def forward(self, x):
...
+
+    def to_model(self):
+        """
+        Merge the sub linear layers into one
+
+        Returns:
+            `nn.Linear` layer
+        """
+        dtype = self.split_linear_layers[0].weight.dtype
+        layer = nn.Linear(self.in_features, self.out_features, bias=False if self.bias is None else True)
```
Comment (katalinic-gc) [inline optimum/graphcore/modeling_utils.py:654]: nit: `bias=self.bias is not None`

**Review by katalinic-gc — APPROVED** [review]

File: optimum/graphcore/ipu_configuration.py [hunk 11]
```diff
@@ -161,14 +194,18 @@ def __get__(self, obj, objtype=None):
...
 
     # Create a mapping of attributes to their list of validation functions
     attribute_validators = defaultdict(list)
+    _layers_per_ipu = ManagedAttribute("layers_per_ipu")
+    _ipus_per_replica = ManagedAttribute("ipus_per_replica")
+    _matmul_proportion = ManagedAttribute("matmul_proportion")
+    _projection_serialization_factor = ManagedAttribute("projection_serialization_factor")
+    _serialized_projection_splits_per_ipu = ManagedAttribute("serialized_projection_splits_per_ipu")
+    _embedding_serialization_factor = ManagedAttribute("embedding_serialization_factor")
+    _serialized_embedding_splits_per_ipu = ManagedAttribute("serialized_embedding_splits_per_ipu")
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:206]: Why the change in location?

File: optimum/graphcore/ipu_configuration.py [hunk 12]
```diff
@@ -111,10 +111,43 @@ class IPUConfig(BaseConfig):
             Same as `matmul_proportion` for inference only.
         enable_half_partials (`bool`, *optional*, defaults to `True`):
             If `True`, sets the data type of partial results for matrix multiplication and convolution operators to float16.
-        embedding_serialization_factor (`int`, *optional*, defaults to 1):
+        projection_serialization_factor (`int`, *optional*, defaults to 1 if `serialized_projection_splits_per_ipu` is `None`):
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:114]: Would it make sense to leave the embedding option at the top as it appears "first" in the models this will be used in?

File: optimum/graphcore/ipu_configuration.py [hunk 13]
```diff
@@ -317,15 +401,22 @@ def check_and_set_replication_factor(attr_name, attr, default=False):
         if "enable_half_first_order_momentum" in kwargs:
             warnings.warn('The "enable_half_first_order_momentum" parameter is deprecated')
 
+        self.gradient_accumulation_steps = gradient_accumulation_steps
+        self.device_iterations = device_iterations
+        self.inference_device_iterations = inference_device_iterations
+        self.optimizer_state_offchip = optimizer_state_offchip
+        self.replicated_tensor_sharding = replicated_tensor_sharding
+        self.auto_loss_scaling = auto_loss_scaling
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:409]: Why the change in location?

File: optimum/graphcore/ipu_configuration.py [hunk 14]
```diff
@@ -419,20 +509,51 @@ def validate_ipu_config(self):
                 matmul_proportion_mode_str = self._get_managed_attr_mode_name("matmul_proportion")
                 raise IncompatibleIPUConfigError(
                     f"{matmul_proportion_mode_str}={self._matmul_proportion} should use the"
-                    f" same number of IPUs as {ipus_per_replica_mode_str}={self._ipus_per_replica}"
+                    f" same number of IPUs as {ipus_per_replica_mode_str}={self._ipus_per_replica}."
                 )
 
-            # layers_per_ipu must have the same length as ipus per replica.
+            # Layers_per_ipu must have the same length as ipus per replica.
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:515]: This is a variable name, it shouldn't be capitalised
```suggestion
            # layers_per_ipu must have the same length as ipus per replica.
```

File: optimum/graphcore/ipu_configuration.py [hunk 15]
```diff
@@ -419,20 +509,51 @@ def validate_ipu_config(self):
...
+                    )
+
+                # Serialized layer splits per ipu pipeline must have the same pipeline length
+                # as the number of ipus per replica
+                if mode_layer_splits_per_ipu and len(mode_layer_splits_per_ipu) != self._ipus_per_replica:
+                    raise ValueError(
+                        f"{mode_layer_splits_per_ipu_str}={mode_layer_splits_per_ipu}"
+                        f" should use the same number of IPUs as {ipus_per_replica_mode_str}={self._ipus_per_replica}."
+                    )
+
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:554]: Could/should these be added as `attribute_validators`?

File: optimum/graphcore/ipu_configuration.py [hunk 16]
```diff
@@ -419,20 +509,51 @@ def validate_ipu_config(self):
...
         self.mode = old_mode
+        return self
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:556]: Why return self? The return value doesn't appear to be used anywhere

File: optimum/graphcore/ipu_configuration.py (hunk 15 above)
Comment (kundaMwiza) [reply optimum/graphcore/ipu_configuration.py:554]: The issue with introducing attribute validators that compare the values of more than one attribute is ensuring that the order of validation is correct. If `*_splits_per_ipu` depends on `*ipus_per_replica` then we must make sure that `*ipus_per_replica` is validated first and it does not also depend on `*_splits_per_ipu`. If there is a circular dependency the only way to resolve attribute value conflicts would be to construct another `IPUConfig`

File: optimum/graphcore/ipu_configuration.py (hunk 16 above)
Comment (kundaMwiza) [reply optimum/graphcore/ipu_configuration.py:556]: I added that to support chaining method calls if `validate_ipu_config` is called elsewhere e.g. trainer.train() { .. ipu_config._validate_ipu_config() }

File: optimum/graphcore/ipu_configuration.py (hunk 13 above)
Comment (kundaMwiza) [reply optimum/graphcore/ipu_configuration.py:409]: Do you mean why does the order not match the docstring/ `__init__` parameter order? I placed them together because they don't have any special initialisation

Comment (HMellor) [reply optimum/graphcore/ipu_configuration.py:409]: I was just wondering why these options were moved to this location

File: tests/test_modeling_utils.py [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+#             options.deviceIterations(len(serialized_embedding_splits_per_ipu))
+
+#             other_evaluator = PoptorchEvaluator(
+#                 serialized_embedding.parallelize(serialized_embedding_splits_per_ipu),
+#                 options,
+#             )
+#         else:
+#             other_evaluator = PytorchEvaluator(serialized_embedding)
+
+#         return evaluator, other_evaluator, dataset
```
[the remark this answers is not in the document]
Comment (kundaMwiza) [reply tests/test_modeling_utils.py:165]: Another PR is needed to refactor `SerializedEmbedding`to use the to/from_model design. I can remove the code

File: optimum/graphcore/ipu_configuration.py [hunk 18]
```diff
@@ -111,10 +111,44 @@ class IPUConfig(BaseConfig):
...
+            serialize the projection layer into a set of individual linear layers that can be optionally placed on different IPUs.
+            Nothing happens if `projection_serialization_factor = 1`. If `projection_serialization_factor > 1`,
+            the `torch.nn.Linear` layer is replaced by a `optimum.graphcore.modeling_utils.SplitProjection` layer
+            if `serialized_projection_splits_per_ipu` is provided and the linear layer's weights are not tied to another layer.
+            Otherwise it is replaced by a `optimum.graphcore.modeling_utils.SerializedLinear` layer.
+            Note: only one of `projection_serialization_factor` or `serialized_projection_splits_per_ipu` should be set.
+        inference_projection_serialization_factor (`int`, *optional*, defaults to 1 if `inference_serialized_projection_splits_per_ipu` is `None`):
+            Same as `projection_serialization_factor` but for inference only.
+        serialized_projection_splits_per_ipu (`List[int]`, *optional*, defaults to None):
+            Analogoues to `serialized_embedding_splits_per_ipu`.
```
Comment (jimypbr) [inline optimum/graphcore/ipu_configuration.py:148]: spelling

**Review by jimypbr — APPROVED** [review]
LGTM
